sklearn/base.py

```diff
@@ -1212,7 +1338,10 @@ def wrapper(estimator, *args, **kwargs):
                     prefer_skip_nested_validation or global_skip_validation
                 )
             ):
-                return fit_method(estimator, *args, **kwargs)
+                try:
+                    return fit_method(estimator, *args, **kwargs)
+                finally:
+                    estimator._eval_callbacks_on_fit_end()
```

Comment:
this is becoming larger than just a validation wrapper. We can simplify debugging and magic by having a `BaseEstimator.fit` which calls `self._fit(...)` and does all the common stuff before and after. That seems a lot better to understand and debug.

Reply:
The motivation when I introduced `_fit_context` was not only for validation but to have a generic context manager to handle everything we need to do before and after fit. That's why I gave it this generic name.

Although having a consistent framework where we'd have a `BaseEstimator.fit` and every child estimator implements `_fit` is appealing, I think it goes far beyond the scope of this PR and requires rewriting a lot of estimators.

Reply:
`BaseEstimator` doesn't implement `fit` cause we don't generally have methods which raise `NotImplementedError`. They're simply not there. But now that we have all this work, we can certainly have it in `BaseEstimator`, and children only implement a `__sklearn_fit__` kind of method instead.

Reply:
I still think it's outside the scope of this PR. Using the existing context manager is just 1 line addition whereas implementing `__sklearn_fit__` means countless PRs :)